### PR TITLE
Add support to browser for gaiahub and keyfiles

### PR DESCRIPTION
**Pull request by kantai**
Okay -- here's some instructions for getting this running with blockstack-core, browser, and blockstack todos using _docker containers_:

Get these repos:
```
git clone https://github.com/blockstack/blockstack-browser -b feature/multi-reader-storage
git clone https://github.com/blockstack/blockstack-core -b feature/multi-reader-storage
git clone https://github.com/kantai/blockstack-todos -b feature/multi-reader-storage
```

Now -- build your containers:
```
cd blockstack-browser && docker build . -f Dockerfile.keyfile.sources --tag browser:multi-reader-storage
cd ../blockstack-core && docker build . -f integration_tests/deployment/docker/Dockerfile.tests --tag blockstack-core:multi-reader-storage
cd ../blockstack-todos && docker build . --tag todos:multi-reader-storage
```

Now, run your containers:

```bash
docker run -dt --name blockstack-core-multi-reader -p 6270:6270 -e BLOCKSTACK_TEST_CLIENT_RPC_PORT=6270 -e BLOCKSTACK_TEST_CLIENT_BIND=0.0.0.0 blockstack-core:multi-reader-storage blockstack-test-scenario --interactive 2 blockstack_integration_tests.scenarios.portal_test_env
docker run -d --name blockstack-todos-multi-reader -p 5000:5000 -e BSK_TODOS_PORT=5000 -e BSK_TODOS_BINDADDR=0.0.0.0 todos:multi-reader-storage npm run start
docker run -d --name blockstack-browser-multi-reader -p 8888:8888 browser:keyfile blockstack-browser
docker run -d --name blockstack-cors-multi-reader -e CORSPROXY_HOST="0.0.0.0" -p 1337:1337 browser:multi-reader-storage blockstack-cors-proxy
```

Open your browser to `localhost:8888`, make a profile, etc. and use blockstack-todos from `localhost:5000`

This works for me in regtest (and mainnet without a username) -- would be good to get other people testing this out as well.

See (now closed) PRs #739  #808

### CODE REVIEW

File: app/js/utils/api-utils.js [hunk 1]
```diff
@@ -4,9 +4,13 @@ import log4js from 'log4js'
 
 const logger = log4js.getLogger('utils/api-utils.js')
 
-import { uploadProfile, DROPBOX } from '../account/utils'
+import { uploadProfile, DROPBOX, BLOCKSTACK_INC } from '../account/utils'
 import { signProfileForUpload } from './index'
 
+import { keyFileCreate, keyFileParse, keyFileProfileSerialize, keyFileUpdateApps, keyFileMakeDelegationPrivateKeys, keyFileICANNToAppName, getPubkeyHex, datastoreMountOrCreate } from 'blockstack'
+const jsontokens = require('jsontokens')
+const assert = require('assert')
```
Comment (larrysalibra) [inline app/js/utils/api-utils.js]: the `assert` package is for unit testing and isn't intended to be used in shipped code: see https://www.npmjs.com/package/assert

File: app/js/utils/api-utils.js [hunk 2]
```diff
@@ -143,15 +146,20 @@ function profileInsertStorageRoutingInfo(profile, driverName, indexUrl) {
 export function setCoreStorageConfig(api,
   blockchainId = null, profile = null, profileSigningKeypair = null) {
   return new Promise((resolve, reject) => {
+    var driverName = null;
```
Comment (larrysalibra) [inline app/js/utils/api-utils.js]: There's a lot of lint here. can we clean this up?

File: app/js/utils/api-utils.js [hunk 3]
```diff
@@ -205,3 +213,89 @@ export function setCoreStorageConfig(api,
...
+   return new Promise((resolve, reject) => {
+      const session = jsontokens.decodeToken(sessionToken).payload
+      const blockchainId = session.blockchain_id
+      const deviceId = session.device_id
+      const appName = session.app_domain
+      const datastoreId = session.app_user_id
+
+      assert(deviceId, 'Invalid session: missing device ID')
+      assert(appName, 'Invalid session: missing application name')
+      assert(datastoreId, 'Invalid session: missing datastore ID')
```
Comment (larrysalibra) [inline app/js/utils/api-utils.js]: A better way to handle this than using `assert` is to check for the value and if it is missing, log a useful error message and `reject()` the promise.

File: app/js/App.js [hunk 4]
```diff
@@ -60,9 +58,9 @@ class App extends Component {
     super(props)
 
     this.state = {
-      accountCreated: !!this.props.encryptedBackupPhrase,
-      storageConnected: !!this.props.dropboxAccessToken,
-      coreConnected: !!this.props.api.coreAPIPassword,
+      accountCreated: this.props.encryptedBackupPhrase ? true : false,
```
Comment (larrysalibra) [inline app/js/App.js]: These changes introduce lint and break the automated test suite. (`npm run test` tests for lint in files that were previously verified as lint-free)

File: app/js/utils/api-utils.js (hunk 2 above)
Comment (larrysalibra) [reply app/js/utils/api-utils.js]: I'll clean it up during testing.

**Comment by kantai** [comment]
> I made a few style comments.
Thanks! I'll be sure my javascript lints okay in the future.

**Comment by jackzampolin** [comment]
I have this is working for both no username and for users with names.

**Comment by larrysalibra** [comment]
Moving this to backlog - it's on hold for a while.

**Comment by larrysalibra** [comment]
@kantai I'm going to close this pull request since we're going with the apps write directly to gaia hubs and light weight multi-reader storage approaches. It if should stay open, please re-open!

**Pull request closed without merging**